### ISSUE TITLE
Discussion and development of web/wasm version

### ISSUE THREAD

**Issue by falkTX** [DISTRHO/Cardinal#287]
For those unaware, https://cardinal.kx.studio/ is totally a thing now for some reason.
Also https://minicardinal.kx.studio/ for a mini/stripped-down version that loads faster.

There are some issues and things still to do.
Because this is for playground and testing and fun, and not all that serious, let's not open more than 1 ticket for web related things.
Everything web/wasm related should be on this ticket.

Stuff to do or implement:
- [x] audio input
- [ ] touch events
- [x] clipboard (do fake internal one if not possible to expose in browser)
- [x] webmidi
- [ ] host time controls (on menu bar, providing fake but valid BBT values)
- [x] intro message explaining what it all is, what to expect etc
- [x] detect incompatible browser setups and refuse to even load (noscript, no simd, no webgl2)
- [x] build libsndfile and fftw for extra module features
- [ ] minimum gui size (CSS is enough?)
- [x] allow url parameters to tweak page behaviour, also load patches as base64 encoded string
- [x] test loading all modules and disable those with custom non-GLES2 compat code

Known issues:
- [x] Loading zst compressed project file fails to load patch.json
- [x] opening salamonis-mono-regen example patch triggers an exception
- [x] going full screen can mess up with the viewport size and coordinates
- [x] ctrl+a clears all text
- [x] audiofile player is missing regular audio support (only mp3 works)
- [x] audiofile and midifile inside ildaeil fail to show file dialogs
- [x] latest changes made performance in chrome worse?? firefox still fine..
- [x] compressed data assets > 50Mb
- [x] missing versioning to js, wasm and data downloads
(so they are updated automatically when reopening the page as needed)

Stuff to fix or workaround:
- [x] apply high-dpi event offset
- [x] test and implement fractional scaling "high-dpi" (only supports 2x and 3x now)
- [x] prevent ctrl/cmd+ shortcut keys from going to the browser
- [x] prevent browser from zooming the page
- [x] oddities with mouse down and leaving browser window

Undecided:
- [ ] host parameters
- [ ] host cv
- [x] persistent storage (client-side browser DB)
- [x] bundle a few simple LV2 plugins (mda, caps, tap, etc)

**Comment by falkTX** [comment]
Leaving a note to myself, https://github.com/Dreamtonics/juce_emscripten/blob/master/modules/juce_audio_devices/native/emscripten/RtMidi.cpp has ruff but working code for webmidi.
Not that it is complicated to do, but helps to have something to compare against.

**Comment by falkTX** [comment]
high-dpi screen should work now.
tested locally with a mac mini.

if you have such a screen, let me know how it works!

![image](https://user-images.githubusercontent.com/1334853/177776938-d68b01aa-27a8-450c-ab30-eb04cb803d4e.png)

**Comment by john-the-bard** [comment]
The ping pong delay from MSC makes a ton of noise if you load it, this basic patch which should play a single plucked note every second into a short and feedback heavy delay.

NOISE WARNING the output is connected and it immediately clips.

[NOISEWARNING_Pluck.tar.gz](https://github.com/DISTRHO/Cardinal/files/9065541/NOISEWARNING_Pluck.tar.gz)

**Comment by falkTX** [comment]
> Insanity. Wowzers. Would be cool to load patch from url encoded parameter, `https://cardinal.kx.studio/?patch=encoded_blob` [like this](https://falstad.com/circuit/circuitjs.html?ctz=CQAJIlh8nHcVBcLpDRgeKEomQqHgAe3AuHEqkaBDM53guEEKHQkW4QA)

I thought about this, and while I typically dislike such approaches, but I think it makes sense here.
A way to generate such link directly from the page would also be quite cool, so we can:
1. open cardinal web page
2. make a patch showing a specific thing
3. use file -> generate pasteable link (or whatever better name there is)
4. send that link to someone for quickly trying the patch

there are 2 things though that need to be done before we can tackle this:
1. fix opening zst compressed patches
2. allow for versioned caching of wasm data, so page reloads are instant (loading 100Mb of resources everytime to check new patches is just painful)

will add to the todo, thanks for the idea!

**Comment by pauldavisthefirst** [comment]
drag doesn't work to scroll boundaries of view area. click on output "socket", drag upward to reach module that is currently "off-screen" .. drag will break.

**Comment by pauldavisthefirst** [comment]
if scroll is used to zoom, easy to get to a point where cardinal does not occupy entire browser screen, at which point the coordinate determination seems to be messed up. drags will affect things not under the mouse.

**Comment by falkTX** [comment]
> drag doesn't work to scroll boundaries of view area. click on output "socket", drag upward to reach module that is currently "off-screen" .. drag will break.

mentioned on the 1st post as
> oddities with mouse down and leaving browser window



> if scroll is used to zoom, easy to get to a point where cardinal does not occupy entire browser screen, at which point the coordinate determination seems to be messed up. drags will affect things not under the mouse.

known issue, I fixed it already locally but didnt deploy a build for it yet

**Comment by falkTX** [comment]
Small update: fixed up the ildaeil audiofile and midifile plugins, adjusted for more high-dpi screens, and updated view menu to follow official Rack 2.1.2 changes.

With the ildaeil fixes it is now possible to get some MIDI going.
Follow these steps:

1. Add ildaeil module
2. Add ildaeil expander midi output, place it touching the right side of ildaeil
3. In ildaeil, switch to internal plugins
4. Load "MIDI File"
5. Set "Host sync" to off, "Repeat Mode" to on
6. Press "show custom gui" button
7. Pick a MIDI file from your PC

Right-click the expander module for polyphony options.

![image](https://user-images.githubusercontent.com/1334853/177987097-3a84d39b-4393-4fbe-8520-4a777fcfe2ed.png)

**Comment by falkTX** [comment]
Good news, the issue with loading compressed project files is fixed. (or better said, a workaround is in place)
Somehow zstd extracts the tar with incorrect permissions, and then fails to load.
We for 777 for any file to be open, and all is good. The "files" are all virtual anyway, so doesn't matter on the web context.

Also, you can now pass it a patchstorage "slug" to the URL as a way to load that patch on start.
For example https://cardinal.kx.studio/?patchstorage=fern-sequence and https://cardinal.kx.studio/?patchstorage=example-patch-2
The URL parameter is the same that is used on patchstorage.com views, e.g. https://patchstorage.com/example-patch-2/

**Comment by falkTX** [comment]
Another update, this is fun :laughing: 

Thanks to previous work done in https://github.com/DISTRHO/PawPaw I could just repurpose it to build generic opensource libs for wasm. For now that is libogg, libvorbis, flac, opus, sndfile, liblo, fftw and lv2 related libs.
This helps a lot on building opensource plugins in wasm.
Also these extra libs allow to get the extra/proper file support in the audiofile module and brings in the audio-to-cv-pitch module (which required fftw)

I did a few more tweaks and made Cardinal able to load audio plugins from the "filesystem" (JSFX dont really count, they are not binaries).
I have compiled artyfx, die-plugins (imported ardour fx), fomp and mda LV2 plugins, and added them to the site/webapp. Just open up ildaeil and they will be there for the LV2 option.

![image](https://user-images.githubusercontent.com/1334853/178124000-1b6991d5-92e8-4238-85ee-b4f6fdd21389.png)

**Comment by falkTX** [comment]
ok a couple more things now:
- fixed up event key input handling a bit, ctrl+a no longer clears all text
- keyboard is no longer propagated to browser, so stuff like ctrl + scroll works now (some key combos like ctrl+n and ctrl+t cannot be overridden)
- pointer lock is implemented and enabled by default (opt-out in view menu)
- added full screen menu option

**Comment by dromer** [comment]
Going out of full screen returns to the wrong size render (x/y mouse coordinates are still correct, but don't align with the view).

**Comment by falkTX** [comment]
> Going out of full screen returns to the wrong size render (x/y mouse coordinates are still correct, but don't align with the view).

I have some ideas for this, the fullscreen mode typically changes the viewport to render in "pixel perfect" mode by default, while typically contents are scaled instead. When entering fullscreen, we can request how to deal with the viewport and keep things as-is. Will try later

**Comment by falkTX** [comment]
And more news for the web side, Web MIDI is implemented!
Both MIDI input and MIDI output.
Does not work on Firefox, as it is not implemented there. Only Chrome/ium-based browsers work for that.
And if you are using an Ubuntu based system with snaps, alsa-midi access is disabled there so this wont work.

The web audio stuff was rewritten from scratch too, I was using SDL before but it was doing some weird stuff behind the scenes, specially for audio input. Oh yeah, I got audio input to work too, but did not enable yet as the permission request for it would make the page look too suspicious. ("Why does this random page want mic access?")
A button on the menubar to request mic access might be the best, to handle later.
I did try the web audio input, and quality is really awful, so we are not losing much.

For the curious ones, relevant code for this web audio + midi is at https://github.com/DISTRHO/DPF/blob/develop/distrho/src/jackbridge/WebBridge.hpp
This is not using worklets yet, as that has special security considerations and just a pain to setup.
With the current single-threaded approach we do not get the best performance, but in turn it is also possible to just run the html files directly in any compatible browser.

I will try to have the html/js/wasm files built as part of 22.07 release on CI.
There is nothing special about it, just emscripten with `USING_GLES=true` build flag.

**Comment by falkTX** [comment]
Audio input is now enabled.
And MIDI is disabled by default, as I try to optimize things a little big (chrome is behaviour worse after changing to custom written code vs old SDL2 usage, for some reason)

You can enable audio input and MIDI under the engine menu.

Audio input quality is really subpar here.. and forces itself as mono even if we request 2 channels.
You can see the blocky-ness of it in a spectrum, really ugly :(

![image](https://user-images.githubusercontent.com/1334853/178278716-4b31093c-21c8-4fcd-9980-4b7599bfcce5.png)

But, well, feature is enabled now.

**Comment by falkTX** [comment]
A bit more news on the input and processing..
For one, the issue I had with quality seems related to pulseaudio. using chrome with alsa the input quality is much better.

![image](https://user-images.githubusercontent.com/1334853/178317759-4b2fb054-61b5-4c2a-80b6-602f67643378.png)

Chrome cannot do stereo input though, only Firefox does.

Also "fixed" the choppy audio in chrome after the latest updates, turns out it is just tweaking buffer size.
SDL was likely buffering things, not using a real 512 samples as requested.
Default now changed to be 2048 samples, but you can change this in the engine menu.
Might help make the webapp work on slower machines.

**Comment by falkTX** [comment]
There is a mini version now at https://minicardinal.kx.studio/
Contains only core and fundamental modules, extras like carla/ildaeil are also not in there.

The mini version assets have a combined 15Mb size.
It is small enough that Firefox is happy to cache the page, so that's nice :)

Feel free to use and share!

**Comment by falkTX** [comment]
A few more things:

touch events are now enabled, sorta..
because neither Rack or pugl can deal with touch events, they are converted to mouse events in hopefully a way that makes sense.
- only 1 finger used at the moment.
- quick press does a mouse click
- press and hold does mouse right click
- press and move does a mouse click + drag (so if you dont want to trigger a right click menu, move finger after pressing)

moving around the canvas is still an issue. I am considering mapping 2 finger movements to mouse wheel, will need to run some tests..
I managed to make it load on my phone, but since it is a > 7 year old one, its performance is quite terrible. otherwise works.
testing and feedback is appreciated.

the hosted files are now gzip compressed, which makes them smaller and thus faster to load.
we can gzip compress the files and host both compressed and uncompressed, then apply some apache magic to load compressed files if browser supports it.
it is a nice reduction in size!

![image](https://user-images.githubusercontent.com/1334853/182838235-dd6f543e-50c0-451c-ac91-6c2629e777af.png)

I will push the setup for this later.

A final thing added was detection of required browser features before trying to load the web assembly program.
Old browsers cant load Cardinal web, but that is kinda expected, it is nice to have a list of unsupported features.

![image](https://user-images.githubusercontent.com/1334853/182839043-a3269141-aa27-4d31-ad34-96b92154866e.png)

And speaking of unsupported browsers, with this we now know why it does not work on Safari - lack of wasm simd support.
That means it wont work on any iOS device at the moment too..
Apple seems to be pushing in wasm territory though, so perhaps this will be coming soon.
https://platform.uno/blog/the-state-of-webassembly-2021-and-2022/ seems to give a good overview of the current situation.

**Comment by falkTX** [comment]
Oh, and another small change I forgot to mention: when browser tab/window is not visible, cardinal wont spend time rendering the webgl gui anymore. which makes it use less cpu when running in the background.

**Comment by falkTX** [comment]
Good news, with more optimizations and compression, the wasm binary and data downloads are now below <50Mb.

![image](https://user-images.githubusercontent.com/1334853/183314470-888a97cf-6c22-454a-8538-050c63046f72.png)


This is great because, at least in Firefox, downloads with less than 50Mb are allowed to be cached.

![image](https://user-images.githubusercontent.com/1334853/183314591-2caab468-f68d-4633-8d3b-98e48ab1c00f.png)

Additionally I did some tests for clipboard support.
While it is possible to support it, and I verified that it actually works, mapping it to current APIs as used by DPF and Rack requires us to use [asyncify](https://emscripten.org/docs/porting/asyncify.html) which complicates the build and makes it slower. in the end it seems not worth it to pursue.
note that this is only for pasting from system clipboard into the site/webapp, the other way around does not have such strong limitations. we can even fake/force copying out of the webapp by creating a hidden html element, setting its text to the one we want and instruct the browser to copy from it.

so starting today we have a sorta fake clipboard in the webapp, with it having internal data for dealing with copy&paste and allowing to copy out of the webapp as an extra.

**Comment by falkTX** [comment]
We have URL parameters handling now!

Both passing full patch, or a link to some file online.

Use `patch` keyword for patch file in URL, alike https://cardinal.kx.studio/?patch=eyJtb2R1bGVzIjpbeyJpZCI6IDEsInBsdWdpbiI6IkNhcmRpbmFsIiwibW9kZWwiOiJUZXh0RWRpdG9yIiwiZGF0YSI6eyJldGV4dCI6IldoYXQgYSBoYWNrIHdvdyJ9fV0sImNhYmxlcyI6W119Cg==

Use `patchurl` for loading an online patch file, alike https://cardinal.kx.studio/?patchurl=https://patchstorage.com/wp-content/uploads/2022/07/JTB-wasm_house_toshare.vcv

The embed patch in the URL is kinda disappointing, even with zst compressed data we easily get into the limits of how long we can make the URL. It works, but it is very limited..
It always uses base64 encoded string, if there is a better way to pass binary data as string please tell me.

The remote file seems the best approach here. you can host a file on a pastebin like service, and pass the raw contents url as parameter to cardinal web app. Can be quite useful to quickly show some patch to someone.

Have fun!

**Comment by Ulfmusic** [comment]
Hi Could I please have some simple instructions on hosting the wasm version of Cardinal? 
Thanks for a great DAW

**Comment by falkTX** [comment]
Download the artifacts and serve them under a web server. There is no magic needed, these are just static files.
Either grab the 22.07 release files, or the latest build from the actions tab.

**Comment by aolney** [comment]
In a word, amazing :partying_face: 

I'm looking at incorporating this into a tutorial website/book I'm working on. 
Some questions with that in mind:

- Looks like computer keyboard for midi input was not ported over. Staying in-browser and ideally within Cardinal, is there an option besides the Impromptu 12-key module? This type of capability is really useful for beginner patches.
- Can menu-based settings be set from the outside? It would be great to 
    - Detect Chrome and automatically turn off `Lock cursor while dragging` to avoid "Press esc to show your cursor" messages
    - Possibly change scrolling behavior for mobile phones (not getting a bottom scroll bar on Android currently)
    - Possibly change the buffer size for mobile phones to something larger than the default

**Comment by falkTX** [comment]
Cardinal does not do midi-like input via keyboard, never did.
For such changes, you can fork the repo and do the changes as you wish.

**Comment by falkTX** [comment]
Small update: The main https://cardinal.kx.studio/ has been updated to the latest code.
Sadly this brings the size of the compressed assets above 50Mb again, and thus they are not being cached by the browser anymore.
Files are also not being automatically updated (their cache invalidated) so that is a small thing to fix too

**Comment by aolney** [comment]
Thanks - going back in the build history, it looks like wasm builds were commented out in [build 1188](https://github.com/DISTRHO/Cardinal/commit/ac71bffae73f5bc2607078088c04c30b032ff9c1).

The last successful build before that was [build 1182](https://github.com/DISTRHO/Cardinal/commit/af64bc5991a68c5f112f26e9efc29134214ddcfb). 
Would this build be the best build for wasm that is still cache friendly?

FWIW I currently have [build 1153](https://github.com/DISTRHO/Cardinal/commit/6bd74114c720c8a806f0e0c9f76029ed50af199d) deployed for testing/evaluation.
I'm deploying through GitHub Pages, so caching is helping me avoid going over their bandwidth quota.

**Comment by falkTX** [comment]
wasm builds are enabled already. they were failing due to new modules having a lot more code which made the build machine go out of RAM. I have reverted voxglitch to the same version as used in 22.07 (with minor fixes), but we are at the limit of what the free github CI VMs can do..

**Comment by aolney** [comment]
> 

Brotli compression seems to solve the caching problem. On build 1153, I get these results



| File                    | Size on disk |
|-------------------------|:------------:|
| CardinalNative.data     |     89.4     |
| CardinalNative.data.zip |     51.2     |
| CardinalNative.data.br  |     35.8     |

The brotli file at level 11 compression is a 30% reduction over zip and 60% reduction over uncompressed.
If browser cache entries must be < 50mb, then the original file could approach 125mb and still be cached using brotli level 11.
I'll report back after I configure nginx for for brotli and live test.

**Comment by falkTX** [comment]
https://github.com/DISTRHO/Cardinal/blob/main/src/emscripten/patchstorage.php has one of the files with https://github.com/DISTRHO/Cardinal/blob/main/src/emscripten/htaccess being what I use in terms on apache setup.

The patchurl.php file I did not post yet because I need to make it better, this is stuff that ends up running server-side after all. So people more knowledgeable of php than me can potentially do bad things with it.

**Comment by aolney** [comment]
Thanks :) Brotli works great live as expected; I'm using nginx. 

I get caching on Firefox for everything, but for some reason on Chrome `CardinalNative.data` will not cache. It looks like Chrome is refusing to send `if-none-match` in its request. 

Did you ever get Chrome to cache `CardinalNative.data`?

**Comment by falkTX** [comment]
Good question, I never used chrome for daily stuff so I didnt check there. Could be it has a more strict caching requirement.
You can try https://minicardinal.kx.studio/ as comparison, as that will have much lower file sizes. If it is possible to cache, the minicardinal will say.

**Comment by falkTX** [comment]
@aolney btw, because you mention incorporating into a website/book, be aware that due to licensing of many modules artwork and design, Cardinal is strictly a non-commercial project.
You cannot involve anything commercial in regards to Cardinal (be it advertising for revenue, donations, merchandise, selling of services, etc, whatever can make something commercial).

If you want to have commercial services related to a Cardinal build, you need to strip out a considerable amount of modules and all the "component library" stuff from VCV.

**Comment by aolney** [comment]
> @aolney btw, because you mention incorporating into a website/book, be aware that due to licensing of many modules artwork and design, Cardinal is strictly a non-commercial project. You cannot involve anything commercial in regards to Cardinal (be it advertising for revenue, donations, merchandise, selling of services, etc, whatever can make something commercial).
> 
> If you want to have commercial services related to a Cardinal build, you need to strip out a considerable amount of modules and all the "component library" stuff from VCV.

No worries, the book is CC-BY-NC-SA 4.0 ;)

BTW I think I've run this down in Chrome - Chrome [is not caching .data files over some ??? size](https://github.com/emscripten-core/emscripten/issues/12212). That page also discusses some ideas about breaking data into chunks, and [this page](https://emscripten.org/docs/compiling/Deploying-Pages.html#providing-a-quick-second-time-load) discusses manually caching to IndexedDB.

So although Brotli is nicely accelerating things, it does not reduce the size of the .data file enough for Chrome to cache it; I went back to build 1153, where the .data file is 35.7MB with brotli, and it's still too big to cache.

**Comment by falkTX** [comment]
IndexedDB is being phased out by browsers, not something to rely on

did you try the wasm-mini builds? there is a small chance chrome doesnt cache any wasm data at all

**Comment by aolney** [comment]
Yes sorry, the minicardinal files are completely cached, both the .wasm and the .data. So I know that Chrome is capable of caching with the headers I'm using, but it decides not to if the .data file is too big (it always caches the .wasm file in my tests).

**Comment by aolney** [comment]
Found a bug in Bogaudio Analyzer-XL. If you connect Noiz to it, it throws an exception and crashes the page. 

BTW I made a repo for dockerized nginx deployment using brotli and php [here](https://github.com/aolney/cardinal-nginx-deploy).

**Comment by aolney** [comment]
> Found a bug in Bogaudio Analyzer-XL. If you connect Noiz to it, it throws an exception and crashes the page.

It seems the Bogaudio analyzers crash with any input. I assume this is the reason for Sassy Scope? Unfortunately Sassy is a bit too noisy to get clear frequency peaks with all the settings I've tried, though it's fine for showing spectrum for many other uses.

**Comment by falkTX** [comment]
sassy was because we could, it is an imgui-based widget after all and can be quite useful.
I suppose bogaudio doesnt crash on desktop? Does it crash on windows or linux 32bit perhaps? it could be a 32bit compat thing, as wasm target runs in 32bit mode as well.

**Comment by aolney** [comment]
> Does it crash on windows or linux 32bit perhaps? it could be a 32bit compat thing, as wasm target runs in 32bit mode as well.

Apologies for the delayed response; it's crashing on 64bit linux. 

BTW found a new strange bug today that is periodically appearing on Firefox when I try to load a patch from the file menu:

![cardinal-tmp-error](https://user-images.githubusercontent.com/4432084/199320710-02462779-2f93-4315-beb9-5c044076d7bc.png)

If I reload the page it goes away. There's no `/tmp/Cardinal...` on my disk, so I assume it's referring to a virtual filesystem?

**Comment by diimdeep** [comment]
Tip: FF users can increase cache limits in `about:config` e.g. `browser.cache.disk.max_entry_size` > 50mb and `browser.cache.disk.capacity` ~ 1gb

**Comment by aolney** [comment]
> Tip: FF users can increase cache limits in `about:config` e.g. `browser.cache.disk.max_entry_size` > 50mb and `browser.cache.disk.capacity` ~ 1gb

Nice! If only Chrome had a similar option

**Comment by davybjj** [comment]
Hello,
Thank you for your work. I don't know if this is the right place. If the following helps.
I've been using the web version for a while in the office because I can't install it. Since the last build a few days ago, the app no longer wants to open on a Windows10 + Chrome Version 108.0.5359.95 (Official Build) (64 bit). It was working until last weekend. I also tried at home on ubuntu + chrome, that doesn't work either. And also with ubuntu 22.04+ firefox with alsa behind, which crashed the machine. For all practical purposes, here is the message in the html page: This is the windows10+chrome computer: I am not a programmer but the html-console displays error messages.
[cardinal.kx.studio-1670427286423.log](https://github.com/DISTRHO/Cardinal/files/10177555/cardinal.kx.studio-1670427286423.log)

**Comment by falkTX** [comment]
no ideas there, logs are not useful. you tried to access it on work office/space? they might have blocked the domain?

I made the page versioned just a few moment ago, to fix the issue of updates being rolled out partially. now the data package and web assembly will always match

in any case, you can download the release wasm zip files and just open the html file inside, so you dont rely on an online version being there. that way you pick which version to use too.

**Comment by aolney** [comment]
@davybjj in your log file, the last line is this:

`Échec du chargement de la carte source par les Outils de développement : Impossible de charger le contenu de chrome-extension://cfhdojbkjhnklbpkdaibdccddilifddb/browser-polyfill.js.map : Erreur système: net::ERR_FILE_NOT_FOUND`

This makes me wonder if the problem is an extension you've installed. I googled this message, and it seems to be associated with Adblock Plus. However it could also be caused by other extensions.

**Comment by davybjj** [comment]
Thank you, I will try.

Le mer. 7 déc. 2022 à 18:27, Andrew M Olney ***@***.***> a
écrit :

**Comment by davybjj** [comment]
Hello,
I come back to you for my feedback. I cannot run the WASM version. As soon
as the html link is pasted, there is an error message: Exception
thrown:Script error.
In Browser Inspector: Browser does not accept CORS origin from local c:/.
After research no NO-CORS plugin could work to circumvent it. Only one
server will apparently be able to function. I can't try on this workstation.
Disabling browser add-ons did not allow me to run the classic web version.
On the other hand, I can access the mini web version as before. So it
shouldn't be a domain issue.
It even started on my modest android mobile. It's great, she makes the
sound. I guess it's in development.
Thanks again for your work, I follow you.

Le jeu. 8 déc. 2022 à 14:23, Davy Voiturier ***@***.***> a
écrit :

**Comment by falkTX** [comment]
Some small bit of news.

I investigated the pointer lock issue on chrome, seems to be bug in emscripten as reported in https://github.com/emscripten-core/emscripten/issues/9681
I added a workaround for now, so default knob behaviour works under chrome-based browsers now.

Some input keys were not being accepted before, like - = ^ ~ ? characters, should now be corrected.

Updated the web app again just now, not sure I mentioned it in previous posts, but we got stoermelder-packone, MindMeldPatchMaster, Sapphire and SurgeXT modules now.

![image](https://user-images.githubusercontent.com/1334853/207308819-14964383-d0b6-45e5-a7dc-2773f427637f.png)

![image](https://user-images.githubusercontent.com/1334853/207309094-ca2cf222-2626-405f-93e8-317465c143ae.png)

Not from Cardinal/DISTRHO side, but related to the discussion, the just released Firefox 108 now supports WebMIDI :rocket: 
You get this weird scary popup warning, but afterwards it appears to work.

![image](https://user-images.githubusercontent.com/1334853/207308512-5b161bd5-d1c1-48b5-bdc1-c35b89782c74.png)

**Comment by davybjj** [comment]
Merci!

Le mar. 13 déc. 2022 à 12:44, Filipe Coelho ***@***.***> a
écrit :

**Comment by davybjj** [comment]
Good evening,
Thank you, I just updated my compiled linux versions with the new modules.
I will try right away. The web version works like a charm on chrome and
firefox linux. I will try firefox midi soon. When I got your message I left
work, I will try it at the end of the week.
Thank you for your patience and good work.

Le mar. 13 déc. 2022 à 17:42, Filipe Coelho ***@***.***> a
écrit :

**Comment by falkTX** [comment]
Good news, sorta..
If we disable SIMD (at the cost of performance) we can have the web version load in Safari/WebKit.

<img width="1105" alt="image" src="https://user-images.githubusercontent.com/1334853/209824133-64577362-f0e1-4eb1-8b41-bde1c8472ac1.png">

Supposedly this means it would work on iOS too, meaning iPhone and iPad.
Going to try to have no-simd by default on the mini version, for the full thing it seems kinda wasteful of CPU resources.

**Comment by falkTX** [comment]
https://minicardinal.kx.studio/ has been updated to NOT have SIMD optimizations, thus it works under Safari/WebKit now.
Can someone with an iOS device (iPad, iPhone, etc) try it out and report how it works? I dont have such device myself..

**Comment by aolney** [comment]
I confirm it works on a basic ipad (9th generation) :smile: 

Is it possible to put logic in the page that decides which SIMD version to load (e.g. using the agent string)?

**Comment by aolney** [comment]
Seeing an interesting bug related to the 22.12 wasm full version.

In my book, I launch cardinal as a modal iframe. With 22.12, if I add a module from the modal iframe, a few key presses in the search bar causes a crash with this error:

```
Exception thrown:
Uncaught RuntimeError: Aborted(). Build with -sASSERTIONS for more info.
```

However, this error does not occur when using cardinal directly (outside the book).

Has there been any change since version 1208 that would cause this error? Looking at the diff between the respective CardinalNative.htmls does not show any meaningful difference.

**Comment by falkTX** [comment]
A bit of news on the mini version. Instead of being the core and fundamental, I updated it to match the real mini version coming for next release.
So opening https://minicardinal.kx.studio/ you get a very select few modules, instead of just fundamental ones.

There is a wasm-noopt build now, which allows to run the entire thing on webkit browsers (and not just the mini variant).
I still need to find a nice way to have simd by default, with fallback to no-simd build for webkit compat, but wont be done for this next upcoming 23.02 release just yet.

**Comment by falkTX** [comment]
Finally found out a regression with emscripten, they changed the default stack size from 5Mb to 64Kb during an update, that breaks everything... Rack uses 64k stack buffers for decoding vcv files, which easily triggered a stack overflow and everything went to shit after loading a compressed file.

With that out of the way I have pushed the latest changes to both https://cardinal.kx.studio/ and https://minicardinal.kx.studio/
Will write more in detail about what has changed on the web versions for the next release (now planned for July 15), but before that happens I want to ask for general testing on 1 particular thing - persistent storage.

Assuming you are running the latest version (shows 23.05 on the top-right corner of the page) we should now have files saved into browser cache, which are made available on subsquent page reloads.
This means you can actually create stuff, save it and come back to it later.
Files are saved in IDBFS (basically browser cache), and not on the server, so they apply to your local device only.

If you can please launch the web versions, do some little patching, save it and try to resume that work after restarting the browser.
The saved files should be accessible through the "load local file" file menu.

<img width="329" alt="image" src="https://github.com/DISTRHO/Cardinal/assets/1334853/a3b200a3-d843-4f82-9a8e-a92b338ed8b9">

Let me know if that works for you!

**Comment by dromer** [comment]
Just tried this with minicardinal and I could see the `Open local patch` what I had just saved in a previous session.
Very cool!

**Comment by aolney** [comment]
> Let me know if that works for you!

Silly question - are there still builds in the nightlies? I don't see them where  they used to be, so perhaps GitHub has changed the interface or they no longer exist?

**Comment by falkTX** [comment]
they are still there as usual, lots of them

<img width="587" alt="image" src="https://github.com/DISTRHO/Cardinal/assets/1334853/3828aea0-2e57-46b1-aae7-9f2ee365c829">

github is just weird on the middle node graph, breaking the down scrolling 😡

**Comment by aolney** [comment]
> github is just weird on the middle node graph, breaking the down scrolling rage

Thx, it was the scroll that was getting me; I was looking at the left pane instead of the main pane as well.

**Here are my test results for `Open local patch`** 

Using
- Chrome 114.0.5735.133 (Official Build) (64-bit) 
- 23.07 (build 1552)

the local storage seems to not be working completely.

If I

- Save (with name)
- Reload the page

then I can recall a patch using `Open recent`. **However I can not recall the patch if I restart the browser.**

**At no time does `Open local patch` become active**. It is always greyed out.

If I repeat the above with Firefox 115.0 (64-bit), I get the same results **except** I can recall the patch if I restart the browser. Again `Open local patch`  is always greyed out.

**Comment by Vospi** [comment]
I found out about this yesterday, I'm just blown away! Did a full sketch of a song in there and sent a link to all the related communities I know. It could also a powerful, powerful demonstration utility for online or in-person discussions. Instantly accessible powerful sketchpad. I'm so happy this exists! Opened up on my phone browser, too, but can't use it with a touch screen. Mouse even kind of works, but... :) you maybe need a keyboard, I couldn't get much further. Would love love to see the day! My subway time would be enhanced instantly. 

So let me instead list the issues I saw while playing in a browser (Firefox Nightly, mainly).

1) Opening a Stoermelder Strip module just throws an exception immediately.
2) Opening a Mindmeld Eqmaster module just throws an exception immediately.
3) If you record anything into a Lilac Looper, then press Export audio file, the whole tab freezes immediately.
4) If there's an exception thrown, I can't save the project, but I can save the selection. Is this intended, do worse things await in the menu that's blocked by the exception? Or maybe there's a way to dismiss the message?
5) If you use Stoermelder MIrror and save the project, it works fine; if you Save selection and import it back, the connection between modules is now lost. (That might be the nature of the module itself, but I'm just saying.)

Minicardinal suggestions!
I also loved the idea of a minimalist setup and was curious to see the approach to choosing the right modules. I didn't find any seq or clock, but simulated it with an LFO + Surge's lfo. However, I can even fake the reverb cause I have no sample delay as far as I could tell. So that's a suggestion there.

Last one, silly suggestion for the full online version. It's online anyways, can we get Bidoo Antn? :) I'm in love with the idea.

I'm back to toying with the thing and experimenting. Love the work. Thank you so much!

**Comment by Vospi** [comment]
Another little thing I found. Happens in the live version, didn't happen on Windows.

**Alea** module interferes with plugins' positions.

https://github.com/DISTRHO/Cardinal/assets/11078819/220ac1cf-529c-48e6-aee0-1f98bd9b12ec

**Comment by dromer** [comment]
@Vospi this behavior with Alea does happen with the standalone and in Rack2Free for me. Better to report this to the original developer instead I think. Or it's a "feature" and not a "bug".

Anyway not at all specific to the web/wasm build of Cardinal.

**Comment by falkTX** [comment]
https://cardinal.kx.studio/live is updated to latest 24.04 release

I fixed the MindMeld modules that caused issues, they wanted to do multi-threaded things which are not allowed in our wasm setup, fixed by [using web timers with emscripten APIs](https://github.com/CardinalModules/MindMeldModular/commit/8136f0c941b41c57af70a5ebb5839f0086518b2b)

That now allows to load the salamonis-mono-regen example patch, though without a fake/internal transport mechanism the patch is not so useful as-is.

But no crashes is always good, and we have a fully functional EQMaster module on the web now

**Comment by Vospi** [comment]
> https://cardinal.kx.studio/live is updated to latest 24.04 release
> 
> I fixed the MindMeld modules that caused issues, they wanted to do multi-threaded things which are not allowed in our wasm setup, fixed by [using web timers with emscripten APIs](https://github.com/CardinalModules/MindMeldModular/commit/8136f0c941b41c57af70a5ebb5839f0086518b2b)
> 
> That now allows to load the salamonis-mono-regen example patch, though without a fake/internal transport mechanism the patch is not so useful as-is.
> 
> But no crashes is always good, and we have a fully functional EQMaster module on the web now

Love this! Saw this commit recently. Thank you very much! Using wasm all the time.

**Comment by henrebotha** [comment]
Just tried this for the first time. Bless WASM. It's unusably zoomed-out for me, though.

![image](https://github.com/DISTRHO/Cardinal/assets/5593874/c95763ce-7087-4a5b-b8b3-4548bb026c87)

macOS 14.5, Firefox 126.0. Running on a 2020 Macbook Pro with Core i5/Iris Plus.

**Comment by Vospi** [comment]
Big problem: Surge XT Filter seems to be bugged out.
I attach two patches, one with, one without the filter. The patch with the filter, no settings touched, just goes insane instantly (in terms of output).
[biset stereo oscillator random.zip](https://github.com/DISTRHO/Cardinal/files/15499927/biset.stereo.oscillator.random.zip)

**Comment by rl2939** [comment]
Selecting the stoermelder STRIP module:
- Does not place the module in the grid
- Throws an uncaught exception

**Comment by Vospi** [comment]
> Selecting the stoermelder STRIP module:
> 
>     * Does not place the module in the grid
> 
>     * Throws an uncaught exception

Yeah I've reported that in February right above:
https://github.com/DISTRHO/Cardinal/issues/287#issuecomment-1933487384
It seems like "hacky" modules like that do something that wasm version doesn't like.

**Comment by Vospi** [comment]
I forgot about the Surge Filters issue, so here I am again, this time it's a screenshot. The patch isn't needed, it's really not working anymore (but it used to!..) even in the most basic setups. Checked in two browsers, too (Chrome and Firefox).

![Cardinal - _vospi - lemon bjnition sequencer v1 vcv - Google Chrome 2024-11-10 14 35 46](https://github.com/user-attachments/assets/79d4793b-3f08-40fd-bc60-e237f64eb527)
